fs/fuse_apps/fuse_block_device.c

```diff
@@ -0,0 +1,207 @@
+        }
+    } else {
+        res = -ENOENT;
+    }
+
+    return res;
+}
+
+static int block_readdir(const char *path, void *buf, fuse_fill_dir_t filler,
+        off_t offset, struct fuse_file_info *fi,
+        enum fuse_readdir_flags flags) {
+    (void) offset;
+    (void) fi;
+    (void) flags;
+
+    if (strcmp(path, "/") != 0) {
+        return -ENOENT;
+    }
+
+    filler(buf, ".", NULL, 0, 0);
+    filler(buf, "..", NULL, 0, 0);
+    filler(buf, options.filename, NULL, 0, 0);
+
+    return 0;
+}
+
+static int block_open(const char *path, struct fuse_file_info *fi) {
+    int fd;
+
+    if (strcmp(path+1, options.filename) != 0) {
+        return -ENOENT;
+    }
+
+    if ((fi->flags & O_ACCMODE) != O_RDONLY) {
+        return -EACCES;
+    }
+
+    fd = open(options.blockdev, fi->flags);
+    if (fd == -1) {
+        return -errno;
+    }
+    fi->fh = fd;
+
+    return 0;
+}
+
+static int block_release(const char *path, struct fuse_file_info *fi) {
+    (void) path;
+    close(fi->fh);
+    return 0;
+}
+
+static int block_read(const char *path, char *buf, size_t size, off_t offset,
+        struct fuse_file_info *fi) {
+    int fd, res;
+
+    if (strcmp(path+1, options.filename) != 0) {
+        return -ENOENT;
+    }
+
+    if (fi == NULL) {
+        fd = open(path, O_RDONLY);
+    } else {
+        fd = fi->fh;
+    }
+
+    if (fd == -1) {
+        return -errno;
+    }
+
+    res = pread(fd, buf, size, offset);
+    if (res == -1) {
+        res = -errno;
+    }
+
+    if (fi == NULL) {
+        close(fd);
+    }
+    return res;
+}
+
+static struct fuse_operations block_oper = {
+    .getattr    = block_getattr,
+    .readdir    = block_readdir,
+    .open       = block_open,
+    .release    = block_release,
+    .read       = block_read,
+};
+
+static void show_help(const char *progname) {
+    printf("usage: %s [options] <mountpoint>\n\n", progname);
+    printf("File-system specific options:\n"
+           "    --blockdev=<s>      Path to a block device\n"
+           "                        (default: \"/dev/nvme0n1\")\n"
+           "    --filename=<s>      Name of the file exposed in the filesystem\n"
+           "                        (default: \"blockdev\")\n"
+           "\n");
+}
+
+int main(int argc, char *argv[]) {
+    int ret;
+    struct fuse_args args = FUSE_ARGS_INIT(argc, argv);
+
+    /* Set defaults -- we have to use strdup() so that
+       fuse_opt_parse() can free the defaults if other
+       values are specified */
+    options.blockdev = strdup("/dev/nvme0n1");
+    options.filename = strdup("blockdev");
+
+    /* Parse options */
+    if (fuse_opt_parse(&args, &options, option_spec, NULL) == -1) {
+        return 1;
+    }
+
+    /* When --help is specified, first print our own file-system
+       specific help text, then signal fuse_main() to show
+       additional help (by adding `--help` to the options again)
+       without usage: line (by setting argv[0] to the empty
+       string) */
```

Comment:
I totally don't understand this part of the comment.

Reply:
It means it won't execute `show_help()` again. _usage: line_  refers to `printf("usage: ...` part (line 169)

Reply:
I rewrote that part. I think it's better now.